### DISCUSSION TITLE
controlling when q-stepper moves to next step

### DISCUSSION THREAD

**Discussion by michalakis** [quasarframework/quasar#13280] [General - Components / Directives / etc]
Hi guys, is there a way for me to run validation, before a step transitions to the next one? as far as I can tell from the documentation, once the user clicks on the q-stepper navigation, the transition is gonna happen whether you want it or not

what I want is to be able to prevent the transition when a user clicks on a step, run validation, and only if it passes, allow the stepper to proceed to the next step 

is this doable? as far as I can tell from the documentation, its not, but I thought I'd ask just in case

**Comment by pdanpdan** [comment]
Use QStepperNavigation - https://quasar.dev/vue-components/stepper#header-navigation

**Reply by michalakis** [reply]
> Use QStepperNavigation - https://quasar.dev/vue-components/stepper#header-navigation

yes I am using the stepper navigation, and as far as I can tell, I can't control if the step should be navigated to or not, unless there is something in the documentation that escapes me

**Comment by pdanpdan** [comment]
You have a button with action under your control

**Reply by michalakis** [reply]
> You have a button with action under your control

I apologize but I dont understan, I think I am missing something

**Comment by MilosPaunovic** [comment]
I think what you're looking for is the `@before-transition` event of the `QStepper`. That's the place to do your validation and decide whether to allow step change.

[https://codepen.io/milospaunovic/pen/bGaXzQJ?editors=1011](https://codepen.io/milospaunovic/pen/bGaXzQJ?editors=1011)

**Reply by michalakis** [reply]
> I think what you're looking for is the `@before-transition` event of the `QStepper`. That's the place to do your validation and decide whether to allow step change.
> 
> https://codepen.io/milospaunovic/pen/bGaXzQJ?editors=1011

I tried using @before-transition, I can run code that starts before the transition takes  place, but the transition still caries on, how do I stop the transition and make it wait for the validation to finish?

edit* I think I see what you did there, you set the step value equal to the old value?

**Reply by MilosPaunovic** [reply]
You assign the old value to property holding your step number if the validation fails to prevent changing step.

**Reply by michalakis** [reply]
> I think what you're looking for is the `@before-transition` event of the `QStepper`. That's the place to do your validation and decide whether to allow step change.
> 
> https://codepen.io/milospaunovic/pen/bGaXzQJ?editors=1011

nope, did it, it still goes to the next step without waiting for the validation

**Reply by michalakis** [reply]
> You assign the old value to property holding your step number if the validation fails to prevent changing step.

so basically, it goes to the next step, and if validation fails, take them to previous step? that has bad optics, and creates an infinite loop

am I missing something here?

**Reply by MilosPaunovic** [reply]
You are right, I missed something. Sorry 'bout that.

**Reply by michalakis** [reply]
> You are right, I missed something. Sorry 'bout that.

no worries, I do appreciate you taking the time to help me

**Comment by rfox12** [comment]
I found that you can pull this off by using a writable computed value (one with a setter).  Then only allow the update if valid....
```
    const innerStep = ref(1);
    const step = computed({
      get() {
        return innerStep.value;
      },
      set(newValue: number) {
        if (innerStep.value !== 1 || validateFields()) { // only move away from step 1 if valid
          innerStep.value = newValue;
        }
      },
    });
```
Then just use `step` as your model for QStepper.
